Source/GameBaseFramework/Private/Characters/Components/GBFAbilityInputBufferComponent.cpp

```diff
@@ -0,0 +1,212 @@
+    TryToTriggerAbility();
+    Reset();
+}
+
+void UGBFAbilityInputBufferComponent::Reset()
+{
+    TriggeredTags.Reset();
+    InputTagsToCheck.Reset();
+    BindHandles.Reset();
+}
+
+void UGBFAbilityInputBufferComponent::BindActions()
+{
+    if ( InputTagsToCheck.IsEmpty() )
+    {
+        return;
+    }
+
+    auto * pawn = GetPawn< APawn >();
+    if ( pawn == nullptr )
+    {
+        return;
+    }
+
+    const auto * hero_component = UGBFHeroComponent::FindHeroComponent( pawn );
+    if ( hero_component == nullptr )
+    {
+        return;
+    }
+
+    for ( auto & input_config : hero_component->GetBoundActionsByInputconfig() )
+    {
+        auto * input_component = pawn->FindComponentByClass< UGBFInputComponent >();
+        if ( input_component == nullptr )
+        {
+            continue;
+        }
+        for ( auto & tag : InputTagsToCheck )
+        {
+            if ( const auto * input_action = input_config.Key->FindAbilityInputActionForTag( tag ) )
+            {
+                BindHandles.Add( input_component->BindAction( input_action, ETriggerEvent::Triggered, this, &ThisClass::AbilityInputTagPressed, tag ).GetHandle() );
+            }
+        }
+    }
+}
+
+void UGBFAbilityInputBufferComponent::RemoveBinds()
+{
+    auto * pawn = GetPawn< APawn >();
+    if ( pawn == nullptr )
+    {
+        return;
+    }
+
+    if ( const auto * hero_component = UGBFHeroComponent::FindHeroComponent( pawn ) )
+    {
+        if ( auto * input_component = pawn->FindComponentByClass< UGBFInputComponent >() )
+        {
+            for ( auto & handle : BindHandles )
+            {
+                input_component->RemoveBindingByHandle( handle );
+            }
+        }
+    }
+}
+
+void UGBFAbilityInputBufferComponent::AbilityInputTagPressed( FGameplayTag input_tag )
+{
+    TriggeredTags.Add( input_tag );
+}
+
+bool UGBFAbilityInputBufferComponent::TryToTriggerAbility()
+{
+    if ( TriggeredTags.IsEmpty() )
+    {
+        return false;
+    }
+
+    auto * pawn = GetPawn< APawn >();
+    if ( pawn == nullptr )
+    {
+        return false;
+    }
+
+    const auto * pawn_ext_comp = UGBFPawnExtensionComponent::FindPawnExtensionComponent( pawn );
+    if ( pawn_ext_comp == nullptr )
+    {
+        return false;
+    }
+
+    if ( auto * asc = pawn_ext_comp->GetGBFAbilitySystemComponent() )
+    {
+        for ( auto & tagged_ability_tag : InputTagsToCheck )
+        {
+            auto * tagged_ability = asc->FindAbilityClassWithInputTag( tagged_ability_tag );
+            asc->CancelAbility( tagged_ability );
+        }
+
+        // Try to activate ability in priority order
+        FGameplayTag tag = TryToGetInputTagWithPriority();
+
+        while ( tag.IsValid() )
+        {
+            if ( auto * ability = asc->FindAbilityClassWithInputTag( tag ) )
+            {
+                asc->CancelAbility( ability );
+                bool activated = asc->TryActivateAbilityByClass( ability->GetClass() );
```

Comment:
auto, or even move it into the if statement

Reply:
Done